libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinGradleBuildServices.kt

```diff
@@ -34,36 +32,69 @@ internal class KotlinGradleBuildServices private constructor(
         val ALREADY_INITIALIZED_MESSAGE = "$CLASS_NAME is already initialized"
 
         @field:Volatile
-        private var instance: KotlinGradleBuildServices? = null
+        internal var instance: KotlinGradleBuildServices? = null
+
+//        @JvmStatic
+//        @Synchronized
+//        fun getInstance(gradle: Gradle): KotlinGradleBuildServices {
+//            val log = Logging.getLogger(KotlinGradleBuildServices::class.java)
+//
+//            if (instance != null) {
+//                log.kotlinDebug(ALREADY_INITIALIZED_MESSAGE)
+//                return instance!!
+//            }
+//
+//            val services = KotlinGradleBuildServices(gradle)
+//            instance = services
+//            if (!isGradleVersionAtLeast(6,1)) {
+//                gradle.addBuildListener(services)
+//                log.kotlinDebug(INIT_MESSAGE)
+//            } else {
+//                BuildEventsListenerRegistry.
+//            }
+//
+//            services.buildStarted()
+//            return services
+//        }
+
 
         @JvmStatic
         @Synchronized
-        fun getInstance(gradle: Gradle): KotlinGradleBuildServices {
+        fun getInstance(project: Project, listenerRegistryHolder: BuildEventsListenerRegistryHolder): KotlinGradleBuildServices {
             val log = Logging.getLogger(KotlinGradleBuildServices::class.java)
+            val kotlinGradleListenerProvider: org.gradle.api.provider.Provider<KotlinGradleBuildListener> = project.provider {
+                KotlinGradleBuildListener(KotlinGradleFinishBuildHandler())
+            }
 
             if (instance != null) {
                 log.kotlinDebug(ALREADY_INITIALIZED_MESSAGE)
                 return instance!!
             }
 
+            val gradle = project.gradle
             val services = KotlinGradleBuildServices(gradle)
-            gradle.addBuildListener(services)
-            instance = services
-            log.kotlinDebug(INIT_MESSAGE)
+            if (isConfigurationCacheAvailable(gradle)) {
```

Comment:
:x: the check should be `gradleVersion >= 6.1` instead where `BuildEventsListenerRegistry` is available

Reply:
New listener is not ready for configuration cache